src/net/p2p.rs

```diff
@@ -116,103 +126,104 @@ struct NetworkHandleServer {
 }
 
 impl NetworkHandle {
-    pub fn new(chan_in: SyncSender<NetworkRequest>, chan_out: Receiver<Result<Option<ReplyHandleP2P>, net_error>>) -> NetworkHandle {
+    pub fn new(chan_in: SyncSender<NetworkRequest>, chan_out: Receiver<Result<Option<ReplyHandleP2P>, net_error>>, result_bufsz: usize) -> NetworkHandle {
         NetworkHandle {
             chan_in: chan_in,
-            chan_out: chan_out
+            chan_out: chan_out,
+            result_buf: VecDeque::new(),
+            result_bufsz: result_bufsz
         }
     }
 
-    /// Connect to a remote peer 
-    pub fn connect_peer(&mut self, neighbor_key: &NeighborKey) -> Result<(), net_error> {
-        let req = NetworkRequest {
-            neighbors: vec![(*neighbor_key).clone()],
-            message: None,
-            expect_reply: false,
-            ttl: 0,
-            connect: true,
-        };
+    fn send_request(&mut self, req: NetworkRequest) -> Result<(), net_error> {
         self.chan_in.send(req).map_err(|_e| net_error::InvalidHandle)?;
-        let res = self.chan_out.recv().map_err(|_e| net_error::InvalidHandle)?;
-        match res {
-            Ok(_) => Ok(()),
-            Err(e) => Err(e)
+
+        if self.result_bufsz > 0 {
+            while self.result_buf.len() < self.result_bufsz {
+                match self.chan_out.try_recv() {
+                    Ok(_res) => {
+                        self.result_buf.push_back(Ok(()));
+                    },
+                    Err(TryRecvError::Empty) => {
+                        break;
+                    },
+                    Err(TryRecvError::Disconnected) => {
+                        return Err(net_error::InvalidHandle);
+                    }
+                }
+            }
+            Ok(())
+        }
+        else {
+            match self.chan_out.recv().map_err(|_e| net_error::InvalidHandle)? {
+                Ok(_) => Ok(()),
+                Err(e) => Err(e)
+            }
         }
     }
 
+    pub fn next_result(&mut self) -> Option<Result<(), net_error>> {
+        self.result_buf.pop_front()
```

Comment:
What is the result_buf used for?

Reply:
It's meant to buffer up replies from the `NetworkHandleServer` instance paired with this `NetworkHandle`.  The p2p network thread can encounter errors when handling a request from the relayer, and the `result_buf` provides a way for the thread with the `NetworkHandle` to get back error messages from the p2p thread.  Right now, the relayer will just drain this buffer and log error messages.

Right now, there's a small bug at line 145 -- the code should be deconstructing `Ok(_res)` and logging the `Err()` case (right now, it just pushes `Ok(())` unconditionally). Will fix.

Reply:
Fixed in 6d83def